sources/APU/DSP/DSP.cpp

```
@@ -257,350 +256,545 @@ namespace ComSquare::APU::DSP
 			break;
 		case 0x7C:
 			for (int i = 0; i < 8; i++)
-				this->_channels[i].endx |= data << i;
+				this->_voices[i].endx |= data << i;
 			break;
 		case 0x0D:
-			this->_registers.efb = data;
+			this->_echo.feedback = data;
 			break;
 		case 0x1D:
-			this->_registers.unused = data;
+			this->_master.unused = data;
 			break;
 		case 0x2D:
 			for (int i = 0; i < 8; i++)
-				this->_channels[i].pmon |= data << i;
+				this->_voices[i].pmon |= data << i;
 			break;
 		case 0x3D:
 			for (int i = 0; i < 8; i++)
-				this->_channels[i].non |= data << i;
+				this->_voices[i].non |= data << i;
 			break;
 		case 0x4D:
 			for (int i = 0; i < 8; i++)
-				this->_channels[i].eon |= data << i;
+				this->_voices[i].eon |= data << i;
 			break;
 		case 0x5D:
-			this->_registers.dir = data;
+			this->_brr.offset = data;
 			break;
 		case 0x6D:
-			this->_registers.esa = data;
+			this->_echo.data = data;
 			break;
 		case 0x7D:
-			this->_registers.edl = data;
+			this->_echo.delay = data;
 			break;
 		case 0x0F:
-			this->_channels[0].coeff = data;
+			this->_echo.FIR[0] = data;
 			break;
 		case 0x1F:
-			this->_channels[1].coeff = data;
+			this->_echo.FIR[1] = data;
 			break;
 		case 0x2F:
-			this->_channels[2].coeff = data;
+			this->_echo.FIR[2] = data;
 			break;
 		case 0x3F:
-			this->_channels[3].coeff = data;
+			this->_echo.FIR[3] = data;
 			break;
 		case 0x4F:
-			this->_channels[4].coeff = data;
+			this->_echo.FIR[4] = data;
 			break;
 		case 0x5F:
-			this->_channels[5].coeff = data;
+			this->_echo.FIR[5] = data;
 			break;
 		case 0x6F:
-			this->_channels[6].coeff = data;
+			this->_echo.FIR[6] = data;
 			break;
 		case 0x7F:
-			this->_channels[7].coeff = data;
+			this->_echo.FIR[7] = data;
 			break;
 		default:
 			throw InvalidAddress("DSP Registers write", addr);
 		}
 	}
+	uint8_t DSP::_readRAM(uint24_t addr) {
+		switch (addr) {
+			case 0x0000 ... 0x00EF:
+				return this->_map.lock()->Page0.read(addr);
```

If _map has expired, this will dereference nullptr and segfault.

---

This should throw a runtime exception & not an InvalidAddress (InvalidAdress will open the debugger if throwned)